scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,352 @@
+use crate::cdc_types::{GenerationTimestamp, StreamID};
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoint {
+    pub timestamp: Duration,
+    pub stream_id: StreamID,
+    pub generation: GenerationTimestamp,
+}
+
+/// Spawns a new task that periodically saves checkpoints.
+/// Receives value of the last checkpoint via channel.
+/// We are using [`tokio::sync::watch`] channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_saving_checkpoints(
+    checkpoint_saver: Arc<dyn CDCCheckpointSaver>,
+    receiver: tokio::sync::watch::Receiver<Checkpoint>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let checkpoint = receiver.borrow().clone();
+
+            checkpoint_saver.save_checkpoint(&checkpoint).await?;
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for saving checkpoints.
+#[async_trait]
+pub trait CDCCheckpointSaver: Send + Sync {
+    /// Saves given checkpoint.
+    async fn save_checkpoint(&self, checkpoint: &Checkpoint) -> anyhow::Result<()>;
+    /// When a new generation comes into effect, it saves it into the checkpoint table.
+    async fn save_new_generation(&self, generation: &GenerationTimestamp) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for [`CDCCheckpointSaver`] trait.
+pub struct CDCDefaultCheckpointSaver {
+    session: Arc<Session>,
+    checkpoint_table: String,
+    make_checkpoint_stmt: PreparedStatement,
+}
+
+impl CDCDefaultCheckpointSaver {
+    /// Creates new [`CDCDefaultCheckpointSaver`].
+    /// If given `None` value as table name,
+    /// it uses default value `"CDC_CHECKPOINT"`.
+    pub async fn new(
+        session: &Arc<Session>,
```

Comment:
Style tip: if you have a function which takes ownership of a shared pointer - like here, you are cloning it and putting it into a struct - then it's better to take `Arc<T>` by value instead of a reference.

Moreover, sometime this lets you avoid a clone, e.g. the caller creates an `Arc<T>`, passes it by value to the function and no clones are made.